proxy/secret.go

```diff
@@ -30,161 +30,136 @@ func (c *Client) GetSecretList(ctx context.Context, namespace string) ([]types.S
 
 	if err != nil {
-		return nil, fmt.Errorf("cannot connect to OpenFaaS on URL: %s", c.GatewayURL.String())
+		return nil, nil, fmt.Errorf("cannot connect to OpenFaaS on URL: %s", c.GatewayURL.String())
 	}
 
 	res, err := c.doRequest(ctx, getRequest)
 	if err != nil {
-		return nil, fmt.Errorf("cannot connect to OpenFaaS on URL: %s", c.GatewayURL.String())
+		return nil, res, fmt.Errorf("cannot connect to OpenFaaS on URL: %s", c.GatewayURL.String())
 	}
 
 	if res.Body != nil {
 		defer res.Body.Close()
 	}
 
+	bytesOut, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, res, err
+	}
+
 	switch res.StatusCode {
 	case http.StatusOK, http.StatusAccepted:
-
-		bytesOut, err := ioutil.ReadAll(res.Body)
-		if err != nil {
-			return nil, fmt.Errorf("cannot read result from OpenFaaS on URL: %s", c.GatewayURL.String())
-		}
-
 		jsonErr := json.Unmarshal(bytesOut, &results)
 		if jsonErr != nil {
-			return nil, fmt.Errorf("cannot parse result from OpenFaaS on URL: %s\n%s", c.GatewayURL.String(), jsonErr.Error())
+			return nil, res, fmt.Errorf("cannot parse result from OpenFaaS: %s", jsonErr.Error())
 		}
 
-	case http.StatusUnauthorized:
-		return nil, fmt.Errorf("unauthorized access, run \"faas-cli login\" to setup authentication for this server")
-
 	default:
-		bytesOut, err := ioutil.ReadAll(res.Body)
-		if err == nil {
-			return nil, fmt.Errorf("server returned unexpected status code: %d - %s", res.StatusCode, string(bytesOut))
+		httpErr := OpenFaaSError{
+			StatusCode: res.StatusCode,
+			Message:    string(bytesOut),
 		}
+		return nil, res, httpErr
 	}
 
-	return results, nil
+	return results, res, nil
 }
 
 // UpdateSecret update a secret via the OpenFaaS API by name
-func (c *Client) UpdateSecret(ctx context.Context, secret types.Secret) (int, string) {
-	var output string
+func (c *Client) UpdateSecret(ctx context.Context, secret types.Secret) (*http.Response, error) {
 	reqBytes, _ := json.Marshal(&secret)
 
 	putRequest, err := c.newRequest(http.MethodPut, secretEndpoint, bytes.NewBuffer(reqBytes))
-
 	if err != nil {
-		output += fmt.Sprintf("cannot connect to OpenFaaS on URL: %s", c.GatewayURL.String())
-		return http.StatusInternalServerError, output
+		return nil, err
 	}
 
 	res, err := c.doRequest(ctx, putRequest)
 	if err != nil {
-		output += fmt.Sprintf("cannot connect to OpenFaaS on URL: %s", c.GatewayURL.String())
-		return http.StatusInternalServerError, output
+		return res, err
 	}
 
 	if res.Body != nil {
 		defer res.Body.Close()
 	}
+	bytesOut, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return res, err
+	}
 
-	switch res.StatusCode {
-	case http.StatusOK, http.StatusAccepted:
-		output += fmt.Sprintf("Updated: %s\n", res.Status)
-		break
-
-	case http.StatusNotFound:
-		output += fmt.Sprintf("unable to find secret: %s", secret.Name)
-
-	case http.StatusUnauthorized:
-		output += fmt.Sprintf("unauthorized access, run \"faas-cli login\" to setup authentication for this server")
```

Comment:
Same question for this one. Shouldn't this either be: part of the typed openfaas error (less repetition), or moved into the command itself?